lib/remap-lang/src/value/object.rs

```diff
@@ -0,0 +1,327 @@
+                Ok(()),
+            ),
+            (
+                map!["foo": Value::Array(vec![0.into(), 1.into()])],
+                vec![Field(Regular("foo".to_owned())), Index(0)],
+                "baz".into(),
+                map!["foo": Value::Array(vec!["baz".into(), 1.into()])],
+                Ok(()),
+            ),
+            (
+                map!["foo": Value::Array(vec![0.into(), 1.into()])],
+                vec![Field(Regular("foo".to_owned())), Index(1)],
+                "baz".into(),
+                map!["foo": Value::Array(vec![0.into(), "baz".into()])],
+                Ok(()),
+            ),
+        ];
+
+        for (object, segments, value, expect, result) in cases {
+            let mut object: Value = object.into();
+            let expect: Value = expect.into();
+            let value: Value = value;
+            let path = Path::new_unchecked(segments);
+
+            assert_eq!(Object::insert(&mut object, &path, value.clone()), result);
+            assert_eq!(object, expect);
+            assert_eq!(Object::get(&object, &path), Ok(Some(value)));
+        }
+    }
+
+    #[test]
+    fn object_remove() {
+        let cases = vec![
+            (
+                map!["foo": "bar"].into(),
+                vec![Field(Regular("foo".to_owned()))],
+                false,
+                Some(map![].into()),
+            ),
+            (
+                map!["foo": "bar"].into(),
+                vec![Coalesce(vec![
+                    Quoted("foo bar".to_owned()),
+                    Regular("foo".to_owned()),
+                ])],
+                false,
+                Some(map![].into()),
+            ),
+            (
+                map!["foo": "bar", "baz": "qux"].into(),
+                vec![],
+                false,
+                Some(map![].into()),
+            ),
+            (
+                map!["foo": "bar", "baz": "qux"].into(),
+                vec![],
+                true,
+                Some(map![].into()),
+            ),
+            (
+                map!["foo": vec![0]].into(),
+                vec![Field(Regular("foo".to_owned())), Index(0)],
+                false,
+                Some(map!["foo": Value::Array(vec![])].into()),
+            ),
+            (
+                map!["foo": vec![0]].into(),
+                vec![Field(Regular("foo".to_owned())), Index(0)],
+                true,
+                Some(map![].into()),
+            ),
+            (
+                map!["foo": map!["bar baz": vec![0]], "bar": "baz"].into(),
+                vec![
+                    Field(Regular("foo".to_owned())),
+                    Field(Quoted("bar baz".to_owned())),
+                    Index(0),
+                ],
+                false,
+                Some(map!["foo": map!["bar baz": Value::Array(vec![])], "bar": "baz"].into()),
+            ),
+            (
+                map!["foo": map!["bar baz": vec![0]], "bar": "baz"].into(),
```

Comment:
Are you able to do like `map! { "foo": map! { "bar baz": true } }`? I notice the `[]` brackets make me think array. 😅